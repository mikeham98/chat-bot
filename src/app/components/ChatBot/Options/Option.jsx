import React from 'react';
import styles from '../../../../themes/components/messages.scss';
import classnames from "classnames";

export default class Option extends React.Component {
    constructor(props) {
        super(props);
        this.onClickOption = this.onClickOption.bind(this);
    }

    returnClassName() {
        const {selected, disabled} = this.props;
        const classNames = [styles.messageOption];
        if (selected) {
            classNames.push(styles.messageOptionSelected);
        } else if (disabled) {
            classNames.push(styles.messageOptionDisabled);
        }
        return classnames(classNames);
    }

    onClickOption(e) {
        const {onClick, disabled} = this.props;
        if (!disabled) {
            e.stopPropagation();
            onClick();
        }
    }

    render() {
        const {option} = this.props;
        return (
            <div className={this.returnClassName()} onClick={this.onClickOption}>
                {option}
            </div>
        );
    }
}