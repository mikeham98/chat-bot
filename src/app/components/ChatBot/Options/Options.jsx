import React from 'react';
import styles from '../../../../themes/components/messages.scss';
import Option from './Option';
import isPopulatedArray from "../../../util/isPopulatedArray";

export default class Options extends React.PureComponent {
    returnOptions() {
        const {options, selectedOption} = this.props;
        if (isPopulatedArray(options)) {
            return options.map(option => {
                return (
                    <Option
                        key={option.id}
                        disabled={selectedOption}
                        selected={selectedOption === option.id}
                        option={option.option}
                        onClick={() => this.props.onClick(option.id)}
                    />
                );
            });
        }
    }

    render() {
        return (
            <div className={styles.messageOptionWrapper}>
                {this.returnOptions()}
            </div>
        );
    }
}