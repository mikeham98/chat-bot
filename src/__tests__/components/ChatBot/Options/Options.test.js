import React from 'react';
import {shallow} from 'enzyme';
import Options from '../../../../app/components/ChatBot/Options/Options';
import Option from '../../../../app/components/ChatBot/Options/Option';

describe('Options', () => {
    let shallowedWrapper;
    let props = {};
    const wrapper = () => {
        if (!shallowedWrapper) {
            shallowedWrapper = shallow(<Options {...props}/>);
        }
        return shallowedWrapper;
    };

    beforeEach(() => {
        props = {
            options,
            onClick: jest.fn(),
            selectedOption: 2,
        };
        shallowedWrapper = undefined;
    });

    describe('render', () => {
        it('should return 2 Option component with the second Option selected', () => {
            expect(wrapper()).toMatchSnapshot();
        });
        it('should set disabled to false when there is not a selected option', () => {
            props.selectedOption = undefined;
            expect(wrapper()).toMatchSnapshot();
        });
    });

    describe('functionality', () => {
        describe('prop', () => {
            describe('onClick', () => {
                it('should be called on onClick with 1 when the first Option is clicked', () => {
                    expect(props.onClick).not.toHaveBeenCalled();
                    wrapper().find(Option).at(0).props().onClick();
                    expect(props.onClick).toHaveBeenCalledTimes(1);
                    expect(props.onClick).toHaveBeenCalledWith(1);
                });
                it('should be called on onClick with 2 when the first Option is clicked', () => {
                    expect(props.onClick).not.toHaveBeenCalled();
                    wrapper().find(Option).at(1).props().onClick();
                    expect(props.onClick).toHaveBeenCalledTimes(1);
                    expect(props.onClick).toHaveBeenCalledWith(2);
                });
            });
        });
    });
});

const options = [
    {
        "id": 1,
        "option": "Back To The Future"
    },
    {
        "id": 2,
        "option": "Stranger Things"
    }
];