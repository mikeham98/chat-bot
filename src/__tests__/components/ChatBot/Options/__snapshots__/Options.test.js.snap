// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Options render should return 2 Option component with the second Option selected 1`] = `
<div
  className="messageOptionWrapper"
>
  <Option
    disabled={true}
    key="1"
    onClick={[Function]}
    option="Back To The Future"
    selected={false}
  />
  <Option
    disabled={true}
    key="2"
    onClick={[Function]}
    option="Stranger Things"
    selected={true}
  />
</div>
`;

exports[`Options render should set disabled to false when there is not a selected option 1`] = `
<div
  className="messageOptionWrapper"
>
  <Option
    disabled={false}
    key="1"
    onClick={[Function]}
    option="Back To The Future"
    selected={false}
  />
  <Option
    disabled={false}
    key="2"
    onClick={[Function]}
    option="Stranger Things"
    selected={false}
  />
</div>
`;
